1\. **Pandas DataFrame**

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a pair of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every `x` BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [ ]:
# If haven't downloaded it yet, please get the data file with wget
#!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ./data/

1\. Create a Pandas DataFrame reading N rows of the `data/data_000637.txt` dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k (check the documentation).

In [2]:
import pandas as pd 
import numpy as np

In [3]:
file_n = "./data/data_000637.txt" 
N = 12000

df = pd.read_csv(file_n, nrows=N)
df

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0         1     0          123  3869200167        2374        26
1         1     0          124  3869200167        2374        27
2         1     0           63  3869200167        2553        28
3         1     0           64  3869200167        2558        19
4         1     0           64  3869200167        2760        25
...     ...   ...          ...         ...         ...       ...
11995     1     0          139  3869200286        3251         0
11996     1     0           62  3869200286        3246         4
11997     1     0           58  3869200286        3246        11
11998     1     0           61  3869200286        3251        17
11999     1     0           59  3869200286        3248        16

[12000 rows x 6 columns]

2\. Estimate the number of BX in a ORBIT (the value `x`).

*Hint*: check when the BX counter reaches the maximum value before being reset to 0.

In [17]:
max_val = 0
N = 12000

orb = df["ORBIT_CNT"]
i = 83
while(orb[i] == orb[i+1]):
    i += 1
print("The number of BX in ORBIT is ", i)

The number of BX in ORBIT is  127


3\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information) since the beginning of the data acquisition, and convert the new column to a Time Series.

In [11]:
df['ABS_TIME'] = np.arange(len(df))

def time(bx,tdc,abs):
    return 

df['ABS_TIME'] = 


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  ABS_TIME
0         1     0          123  3869200167        2374        26         0
1         1     0          124  3869200167        2374        27         1
2         1     0           63  3869200167        2553        28         2
3         1     0           64  3869200167        2558        19         3
4         1     0           64  3869200167        2760        25         4
...     ...   ...          ...         ...         ...       ...       ...
11995     1     0          139  3869200286        3251         0     11995
11996     1     0           62  3869200286        3246         4     11996
11997     1     0           58  3869200286        3246        11     11997
11998     1     0           61  3869200286        3251        17     11998
11999     1     0           59  3869200286        3248        16     11999

[12000 rows x 7 columns]

4\. Find out the duration of the data taking in hours, minutes and seconds, by using the features of the Time Series. Perform this check reading the whole dataset.

5\. Use the `.groupby()` method to find out the noisy channels, i.e. the TDC channels with most counts (print to screen the top 3 and the corresponding counts)

6\. Count the number of non-empty orbits (i.e. the number of orbits with at least one hit).

7\. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139.

8\. Create two Series (one for each FPGA) that have the TDC channel as index, and the number of counts for the corresponding TDC channel as values.

9\. **Optional:** Create two histograms (one for each FPGA) that show the number of counts for each TDC channel.